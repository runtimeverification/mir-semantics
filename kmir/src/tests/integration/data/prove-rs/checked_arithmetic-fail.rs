fn main() {
    checked_add_i32(1,2);
}

/// If the precondition is not met, the program is not executed (exits cleanly, ex falso quodlibet)
macro_rules! precondition {
    ($pre:expr, $block:expr) => {
        if $pre { $block }
    };
}

fn checked_add_i32(a: i32, b: i32) {

    precondition!(
        a.checked_add(b).is_some(),
         unsafe {
            let result = a.unchecked_add(b);
            assert!(result as i64 == a as i64 + b as i64)
        }
    );
}
