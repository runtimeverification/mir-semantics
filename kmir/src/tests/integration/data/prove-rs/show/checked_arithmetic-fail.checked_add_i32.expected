
┌─ 1 (root, init)
│   #execTerminator ( terminator ( ... kind: terminatorKindCall ( ... func: operandC
│   span: 0
│
│  (64 steps)
├─ 3 (split)
│   #selectBlock ( switchTargets ( ... branches: branch ( 0 , basicBlockIdx ( 4 ) )
┃
┃ (branch)
┣━━┓ subst: .Subst
┃  ┃ constraint:
┃  ┃     ARG_INT1:Int +Int ARG_INT2:Int ==K truncate ( ARG_INT1:Int +Int ARG_INT2:Int , 32 , Signed )
┃  │
┃  ├─ 4
┃  │   #selectBlock ( switchTargets ( ... branches: branch ( 0 , basicBlockIdx ( 4 ) )
┃  │
┃  │  (250 steps)
┃  ├─ 6 (terminal)
┃  │   #EndProgram ~> .K
┃  │
┃  ┊  constraint: true
┃  ┊  subst: ...
┃  └─ 2 (leaf, target, terminal)
┃      #EndProgram ~> .K
┃
┗━━┓ subst: .Subst
   ┃ constraint:
   ┃     notBool truncate ( ARG_INT1:Int +Int ARG_INT2:Int , 32 , Signed ) ==K ARG_INT1:Int +Int ARG_INT2:Int
   │
   ├─ 5
   │   #selectBlock ( switchTargets ( ... branches: branch ( 0 , basicBlockIdx ( 4 ) )
   │
   │  (6 steps)
   └─ 7 (stuck, leaf)
       #execIntrinsic ( IntrinsicFunction ( symbol ( "cold_path" ) ) , .Operands , plac



